modules/vanilla/src/entry-server.ts

```diff
@@ -0,0 +1,57 @@
+import type { IncomingMessage, ServerResponse } from "http";
+import { testData } from "testdata";
+
+type EntryData = Awaited<ReturnType<typeof testData>>[0]
+
+function renderEntry(entry: EntryData) {
+  return `
+    <tr>
+      <td>${entry.id}</td>
+      <td>${entry.name}</td>
+    </tr>
+  `
+}
+
+function renderTable(entries: EntryData[]) {
+  let tableHtml = ""
+  for(let i = 0; i < entries.length; i++) {
+    tableHtml += renderEntry(entries[i])
+  }
+
+  return `
+    <table>
+      <tbody>
+        ${tableHtml}
+      </tbody>
+    </table>
+  `
+}
+
+function renderLayout(html: string) {
+  return `
+    <!DOCTYPE html>
+    <html>
+      <head>
+        <title>Vanilla</title>
+      </head>
+      <body>
+        ${html}
+      </body>
+    </html>
+  `
+}
+
+export async function renderHtml() {
+  const data = await testData()
+  const tableHtmlString = renderTable(data)
+  const html = renderLayout(tableHtmlString)
+  return { html }
+}
+
+export async function buildHandler(){
+  return async function handler(_: IncomingMessage, res: ServerResponse) {
+    const { html } = await renderHtml()
+    res.setHeader("Content-Type", "text/html")
+    res.end(html)
+  }
+}
```

Comment:
This part is something like async/suspense logic

Reply:
It is a sync render. "Suspense logic" first renders a fallback html then hydrates it with the resolved content

Reply:
ok, thanks. Then can you please suggest a good alternative？

Reply:
Ok, hmm.
For me, the logic looks the same as the output from solidjs...

Reply:
You could hydrate a shadowroot. It allows for fallbacks content while waiting for the next chunk of the response to stream in:

```html
<body>
  <div>
    <template shadowrootmode=open>
      <slot>Loading...</slot>
    </template>
    <!-- here some async work can happen -->
    ${renderTable()}
  </div>
</body>

Reply:
Who's ready for yet another new framework?

Reply:
![a screen capture of the movie Back To the Future. Doc, Marty & Jennifer are all in the Delorean, and Doc is wearing futuristic shades. The caption says "frameworks?! where we're going we don't need frameworks!"](https://github.com/eknkc/ssr-benchmark/assets/118266/52e388e0-19d8-4e76-bf35-b2cd766ce4a5)

Reply:
I couldn't get it. I think it's enough. Did you see the bundled result from Marko? It's nearly the same as mine.